state-chain/pallets/cf-broadcast/src/migrations/v3.rs

```diff
@@ -0,0 +1,271 @@
+		for id in AwaitingBroadcast::<T, I>::iter_keys() {
+			assert!(pending_broadcasts.contains(&id));
+		}
+
+		let MigrationVerification { broadcasts_with_failed_broadcasters, broadcast_retry_queue } =
+			<MigrationVerification>::decode(&mut &state[..]).unwrap();
+
+		broadcasts_with_failed_broadcasters.into_iter().for_each(|broadcast_id| {
+			assert!(FailedBroadcasters::<T, I>::contains_key(broadcast_id))
+		});
+
+		let new_retry_queue = BroadcastRetryQueue::<T, I>::get()
+			.into_iter()
+			.map(|attempt| attempt.broadcast_id)
+			.collect::<Vec<_>>();
+
+		broadcast_retry_queue
+			.into_iter()
+			.for_each(|broadcast_id| assert!(new_retry_queue.contains(&broadcast_id)));
+
+		Ok(())
+	}
+}
+
+#[cfg(test)]
+mod migration_tests {
+	use super::*;
+	use crate::mock::*;
+
+	#[test]
+	fn migration_works_v2_to_v3_awaiting_broadcast() {
+		new_test_ext().execute_with(|| {
+			// Insert mock data into old storage
+			for broadcast_id in 1..10 {
+				for attempt_id in 0..=broadcast_id {
+					let broadcast_attempt_id =
+						old::BroadcastAttemptId { broadcast_id, attempt_count: attempt_id };
+					let singing_attempt = old::TransactionSigningAttempt {
+						broadcast_attempt: old::BroadcastAttempt {
+							broadcast_attempt_id,
+							transaction_payload: Default::default(),
+							threshold_signature_payload: Default::default(),
+							transaction_out_id: Default::default(),
+						},
+						nominee: attempt_id as u64,
+					};
+					old::AwaitingBroadcast::<Test, Instance1>::insert(
+						broadcast_attempt_id,
+						singing_attempt,
+					);
+				}
+			}
+
+			// Perform runtime migration.
+			crate::migrations::v3::Migration::<Test, Instance1>::on_runtime_upgrade();
+
+			// Verify data is correctly migrated into new storage.
+			// Only the attempt with the highest attempt count is migrated.
+			for broadcast_id in 1..10 {
+				assert_eq!(
+					AwaitingBroadcast::<Test, Instance1>::get(broadcast_id),
+					Some(TransactionSigningAttempt {
+						broadcast_attempt: BroadcastAttempt {
+							broadcast_id,
+							transaction_payload: Default::default(),
+							threshold_signature_payload: Default::default(),
+							transaction_out_id: Default::default(),
+						},
+						nominee: broadcast_id as u64,
+					})
+				);
+			}
+		});
+	}
+
+	#[test]
+	fn migration_works_v2_to_v3_failed_broadcasters() {
+		new_test_ext().execute_with(|| {
+			// Insert mock data into old storage
+			old::FailedBroadcasters::<Test, Instance1>::insert(
+				1,
+				vec![1u64, 1u64, 2u64, 2u64, 3u64, 3u64, 3u64, 3u64],
+			);
+			old::FailedBroadcasters::<Test, Instance1>::insert(2, vec![1u64, 2u64, 3u64]);
+
+			// Perform runtime migration.
+			crate::migrations::v3::Migration::<Test, Instance1>::on_runtime_upgrade();
+
+			// Verify data is correctly migrated into new storage.
+			let failed_1 = FailedBroadcasters::<Test, Instance1>::get(1);
+			let failed_2 = FailedBroadcasters::<Test, Instance1>::get(2);
+			assert!(failed_1.contains(&1u64));
```

Comment:
we also want to check the len = 3 - could also just do this be doing `vec.into_iter().collect::<BTreeSet<_>>()` and comparing the btreesets, directly